containers/Dockerfile

```
@@ -196,52 +186,117 @@ RUN ln -s ${SPACK_ROOT}/share/spack/docker/entrypoint.bash \
 SHELL ["docker-shell"]
 
 # https://spack.readthedocs.io/en/latest/bootstrapping.html
-# Spack is configured to bootstrap its dependencies lazily by default;
+# Spack is configured to bootstrap its dependencies and repositories lazily by default;
 # i.e. the first time they are needed and can’t be found.
 
+# Bootstrap spack and clone relevant spack-packages repos
+
+# Currently, the version of spack-config informs the initial versions of the spack-packages repos to use.
+# If we wanted to do it manually, add this post-bootstrap
+#   builtin_spack_packages_sha=$(git -C $(spack location --repo builtin) rev-parse "${BUILTIN_SPACK_PACKAGES_REPO_VERSION}}^{}")
+#   access_spack_packages_sha=$(git -C $(spack location --repo access.nri) rev-parse "${ACCESS_SPACK_PACKAGES_REPO_VERSION}}^{}")
+#   spack repo update builtin --commit $builtin_spack_packages_sha
+#   spack repo update access.nri --commit $access_spack_packages_sha
+RUN spack bootstrap now
+
 ################################################################################
 # Install compilers and other common packages
-FROM base-spack AS upstream
+FROM base-spack AS base-upstream
 
 LABEL org.opencontainers.image.source=https://github.com/ACCESS-NRI/build-ci
 
 # Install compilers and low-level, common packages
 COPY ${SOURCE_COMPILERS_SPACK_MANIFEST} ${ENV_COMPILERS_SPACK_MANIFEST}
-RUN <<EOF
+RUN --mount=type=secret,id=oci-username,env=OCI_USERNAME \
+    --mount=type=secret,id=oci-password,env=OCI_PASSWORD \
+    --mount=type=secret,id=oci-url,env=OCI_URL \
+    <<EOF
+if [ -n "$OCI_USERNAME" ] || [ -n "$OCI_PASSWORD" ] || [ -n "$OCI_URL" ]; then
+  REQUESTS_OCI_MIRROR=1
+fi
+
+if [ -n "$REQUESTS_OCI_MIRROR" ]; then
+  spack mirror add \
+    --scope=site \
+    --autopush \
+    --unsigned \
+    --oci-username-variable OCI_USERNAME \
+    --oci-password-variable OCI_PASSWORD \
+    docker_build_buildcache $OCI_URL
+fi
+spack mirror list
+
 echo "Installing compilers"
 spack env activate compilers --create --envfile ${ENV_COMPILERS_SPACK_MANIFEST}
-spack spec
 spack install --deprecated --fail-fast || exit
 spack env deactivate
-yq '.spack.specs[]' "${ENV_COMPILERS_SPACK_MANIFEST}" | while read compiler; do
-  echo "Loading compiler: $compiler"
-  spack load $compiler
-  spack compiler find --scope=user
-done
 
 spack clean --downloads --stage
+
+if [ -n "$REQUESTS_OCI_MIRROR" ]; then
+  spack mirror remove --scope=site docker_build_buildcache
+fi
 EOF
 
+# Conditionally add mirrors based on whether secrets exist
 COPY ${SOURCE_PACKAGES_SPACK_MANIFEST} ${ENV_PACKAGES_SPACK_MANIFEST}
-RUN <<EOF
+RUN --mount=type=secret,id=oci-username,env=OCI_USERNAME \
+    --mount=type=secret,id=oci-password,env=OCI_PASSWORD \
+    --mount=type=secret,id=oci-url,env=OCI_URL \
+    <<EOF
+
+if [ -n "$OCI_USERNAME" ] || [ -n "$OCI_PASSWORD" ] || [ -n "$OCI_URL" ]; then
+  REQUESTS_OCI_MIRROR=1
+fi
+
+if [ -n "$REQUESTS_OCI_MIRROR" ]; then
```

```suggestion
if [ -n "$OCI_USERNAME" ] || [ -n "$OCI_PASSWORD" ] || [ -n "$OCI_URL" ]; then
  REQUESTS_OCI_MIRROR=1
```

Can remove redundant `if` and collapse into a single branch.

---

Done in https://github.com/ACCESS-NRI/build-ci/pull/253/commits/e4ba85db0be4a9b9493cf7581623f9997b9404a5, thanks!